src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -5094,8 +5059,47 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         )
         silentResult match {
           case SilentResultValue(fun1) =>
+            def transformRassoc(applied: Tree): Tree = {
+              val Apply(fn, arg :: Nil) = applied: @unchecked
+              val isByName = fn.symbol.paramLists.head.head.isByNameParam
+              if (isByName) applied
+              else {
+                def isRassocArg(t: Tree) = t.hasAttachment[RightAssociativeArg.type]
+                def checkRassocArg(t: Tree): Boolean = isRassocArg(t) && { t.removeAttachment[RightAssociativeArg.type]; true }
+                def rewriteRightAssoc(arg: Tree): Tree = {
+                  val vsym = context.owner.newValue(freshTermName(nme.RIGHT_ASSOC_OP_PREFIX), arg.pos.focus, FINAL | SYNTHETIC | ARTIFACT)
+                  vsym.setInfo(arg.tpe)
+                  val vdef = atPos(arg.pos) { ValDef(vsym, arg) setType NoType }
+                  context.pendingStabilizers ::= vdef
+                  arg.changeOwner(context.owner -> vsym)
+                  Ident(vsym).setType(singleType(NoPrefix, vsym)).setPos(arg.pos.focus)
+                }
+                def needsRewrite(t: Tree) = !treeInfo.isStableIdentifier(t, allowVolatile = false) && !treeInfo.isExprSafeToInline(t)
+                val rewriter = new Transformer {
+                  override def transform(tree: Tree): Tree = tree match {
+                    //case TypeApply(fn, _) if checkRassocArg(fn) =>
+                    //  if (needsRewrite(fn)) rewriteRightAssoc(tree) else tree
+                    case Apply(fn, Apply(arg, Nil) :: Nil) if checkRassocArg(arg) =>
+                      // tree is adapted empty application arg(), so move the adapted tree
+                      if (needsRewrite(tree)) rewriteRightAssoc(tree) else tree
+                    case Apply(fn, arg :: Nil) if checkRassocArg(arg) =>
+                      if (needsRewrite(arg)) treeCopy.Apply(tree, fn, rewriteRightAssoc(arg) :: Nil) else tree
+                    case _ => super.transform(tree)
+                  }
+                }
+                rewriter.transform(addStabilizers(applied))
+              }
+            }
             val fun2 = if (stableApplication) stabilizeFun(fun1, mode, pt) else fun1
             if (settings.areStatisticsEnabled) statistics.incCounter(typedApplyCount)
+            val needsRewrite = {
+              val isRightAssoc = args.lengthCompare(1) == 0 && tree.hasAttachment[RightAssociative.type]
+              isRightAssoc && !isPastTyper && !mode.inPatternMode
+            }
+            if (needsRewrite) {
+              tree.removeAttachment[RightAssociative.type]
+              args.head.updateAttachment(RightAssociativeArg)
```

Comment:
(Why) is this attachment needed?

Reply:
I think all your questions are related. (I had the same questions but I haven't answered them yet.) I think the problem was drilling into adaptations. The rassoc rewrite is after `try/doTypedApply`. At that point, the incoming arg could be moved anywhere in the tree and wrapped arbitrarily. I disable it to see what breaks, then document the tree shapes it is coping with.

Reply:
Possibly you only see one of the arg, `ApplyImplicitView`, or `Block(_, arg or view)`.